### ISSUE TITLE
ForwardAuth bug

### ISSUE THREAD

**Issue by vd2org** [traefik/traefik#6918]
### Do you want to request a *feature* or report a *bug*?

Bug
### What did you do?

I'm trying to use `Headers` and `ForwardAuth` middlewares together like in config described below. 
I use `Headers` to provide `CORS` headers automatically for my services. After `Headers` I use `ForwardAuth`. When` ForwardAuth` receives positive answer(e.g. `200`) everything works good, but in case when `ForwardAuth` receives negative answer(e.g. `403`) response goes straight to client without `CORS` headers. This looks like in the new version the negative answer breaks the chain and upper middlewares are not triggered.

Thus, ajax request from browser can't get response body and can't know what's wrong with the request.

In 2.1 all worked as expected but broken in 2.2.

### What did you expect to see?

CORS headers when auth respond negative code(403 for example).

### What did you see instead?

Response without CORS headers

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```yaml
version: "3.7"
services:

  traefik:
    image: "traefik:2.1"
#    image: "traefik:2.2"
    labels:
      - traefik.http.middlewares.cors.headers.accessControlAllowMethods=GET,POST,PUT,DELETE,OPTIONS
#      Config for 2.2:
#      - traefik.http.middlewares.cors.headers.accessControlAllowOriginList=http://frontend.local
      - traefik.http.middlewares.cors.headers.accessControlAllowOrigin=origin-list-or-null
      - traefik.http.middlewares.cors.headers.accessControlMaxAge=240
      - traefik.http.middlewares.cors.headers.accessControlAllowHeaders=Set-Cookie,Content-Type,authorization,X-Test-Auth
      - traefik.http.middlewares.cors.headers.accessControlAllowCredentials=true
      - traefik.http.middlewares.cors.headers.addVaryHeader=true
      - traefik.http.middlewares.buff.buffering.maxRequestBodyBytes=5242880
      - traefik.http.middlewares.buff.buffering.memRequestBodyBytes=2097152
      - traefik.http.middlewares.buff.buffering.maxResponseBodyBytes=5242880
      - traefik.http.middlewares.buff.buffering.memResponseBodyBytes=2097152
      - traefik.http.middlewares.buff.buffering.retryExpression=IsNetworkError() && Attempts() < 2
      - traefik.http.middlewares.auth.forwardauth.address=http://backend-auth:8080/int/auth/api/v1/auth
      - traefik.http.middlewares.auth.forwardauth.authResponseHeaders=X-Auth-Data
      - traefik.http.routers.traefik.rule=Host(`traefik.local`)
      - traefik.http.services.traefik.loadbalancer.server.port=8080
    command:
      - --providers.docker=true
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - "127.0.0.1:80:80"
    networks:
      - service
    
  backend-auth:
    build:
      context: ./backend/auth/
      dockerfile: Dockerfile.local
    labels:
      - traefik.http.routers.backend-auth-open.rule=Host(`api.local`) && PathPrefix(`/auth/api/v1/login`)
      - traefik.http.routers.backend-auth-open.middlewares=cors@docker,buff@docker
      - traefik.http.routers.backend-auth-open.service=backend-auth-open
      - traefik.http.services.backend-auth-open.loadbalancer.server.port=8080
      - traefik.http.routers.backend-auth.rule=Host(`api.local`) && PathPrefix(`/auth`)
      - traefik.http.routers.backend-auth.middlewares=cors@docker,buff@docker,auth@docker
      - traefik.http.routers.backend-auth.service=backend-auth
      - traefik.http.services.backend-auth.loadbalancer.server.port=8080
    networks:
      - service

  backend-app:
    privileged: true
    build:
      context: ./backend/app/
      dockerfile: Dockerfile.local
    labels:
      - traefik.http.routers.backend-app.rule=Host(`api.local`) && PathPrefix(`/app`)
      - traefik.http.routers.backend-app.middlewares=cors@docker,buff@docker,auth@docker
      - traefik.http.routers.backend-app.service=backend-app
      - traefik.http.services.backend-app.loadbalancer.server.port=8080
    networks:
      - service
```

**Comment by letian0805** [comment]
I found that this commit 082fb166a24138bf7f09f7519c1cd579558cd8ed caused this bug

**Comment by SantoDE** [comment]
Hello @vd2org,

I'm not sure I can follow. Does it also not work, if you set `accessControlAllowOriginList` to the same value as `accessControlAllowOrigin` before?

**Comment by vd2org** [comment]
> I'm not sure I can follow. Does it also not work, if you set `accessControlAllowOriginList` to the same value as `accessControlAllowOrigin` before?

Hi! `accessControlAllowOriginList` is for 2.2 and `accessControlAllowOrigin` for 2.1.

**Comment by SantoDE** [comment]
Yeah, I know :-) But does it work with the same "value" for 2.2 as on 2.1? Or still not? If not, is it maybe linked to #6785 ?

**Comment by letian0805** [comment]
> Yeah, I know :-) But does it work with the same "value" for 2.2 as on 2.1? Or still not? If not, is it maybe linked to #6785 ?

They are two different issues cased by same commit.
1. origin-list-or-null compatibility issues
2. ForwardAuth failed causing CORS issues

**Comment by letian0805** [comment]
Hi @rtribotte 

There is another CORS issue of in traefik v2.2.
Because of it, i can only use version before commit 082fb166a24138bf7f09f7519c1cd579558cd8ed.
I believe there are more users encounter this problem.
I'm studying how to solve it. Can you also take a look?

Thanks!

**Comment by letian0805** [comment]
@ldez 
Do you have any idea about this issue?

**Comment by letian0805** [comment]
I hacked this issue by setting CORS headers in forwardAuth service.

**Comment by vd2org** [comment]
@letian0805 I guess this is not a real solution. CORS middleware is designed exactly for when I don't want to manage it inside applications.

**Comment by letian0805** [comment]
@vd2org Yes! But it works well before this bug is fixed. I read the codes of CORS and ForwardAuth middlewares again last night, and i think i have found a solution to fix it.

**Comment by ddtmachado** [comment]
Hello @vd2org and @letian0805 

I looked at the code and saw no reason for the `ForwardAuth` middleware to break the upper chain, even so I tried validating it  by debugging with breakpoints and by running it on real environments with a compose file based on your setup. Everything worked as expected without any issues so far.

Here is the compose file used, which is based on yours:

```yaml
version: "3.7"

networks:
  traefik:

services:

  traefik:
#    image: "traefik:2.1"
    image: "traefik:2.2"
    labels:
#     Dashboard
      - "traefik.enable=true"
      - "traefik.http.routers.api.entrypoints=web"
      - "traefik.http.routers.api.rule=Host(`traefik.docker.localhost`)"
      - "traefik.http.routers.api.service=api@internal"

#     Headers
      - traefik.http.middlewares.cors.headers.accessControlAllowMethods=GET,POST,PUT,DELETE,OPTIONS
      # Config for 2.2:
      - traefik.http.middlewares.cors.headers.accessControlAllowOriginList=http://mylaptop
      # Config for 2.1:
      #- traefik.http.middlewares.cors.headers.accessControlAllowOrigin=origin-list-or-null
      - traefik.http.middlewares.cors.headers.accessControlMaxAge=240
      - traefik.http.middlewares.cors.headers.accessControlAllowHeaders=Set-Cookie,Content-Type,authorization,X-Test-Auth
      - traefik.http.middlewares.cors.headers.accessControlAllowCredentials=true
      - traefik.http.middlewares.cors.headers.addVaryHeader=true

#     ForwardAuth
      - traefik.http.middlewares.auth.forwardauth.address=https://postman-echo.com/basic-auth      
    command:
      - --providers.docker=true
      - --entryPoints.web.address=:80
      - --api.dashboard=true
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - "80:80"
    networks:
      - traefik

  whoami-http:
    image: "containous/whoami:latest"
    labels:
      - traefik.http.routers.backend-app.rule=Host(`whoami.docker.localhost`)
      - traefik.http.routers.backend-app.middlewares=cors@docker,auth@docker
      - traefik.http.routers.backend-app.service=backend-app
      - traefik.http.services.backend-app.loadbalancer.server.port=80
    networks:
      - traefik
```

You can just launch it on localhost, then send requests that succeed of fail authentication:

```bash
### Auth fail, with Origin header set
$ curl -v -H "Origin:http://mylaptop" -H "Host:whoami.docker.localhost" 127.0.0.1
* Rebuilt URL to: 127.0.0.1/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host:whoami.docker.localhost
> User-Agent: curl/7.52.1
> Accept: */*
> Origin:http://mylaptop
>
< HTTP/1.1 401 Unauthorized
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Origin: http://mylaptop
< Date: Thu, 15 Oct 2020 20:05:45 GMT
< Set-Cookie: sails.sid=s%3AnvxcrP56Tr4ecPrtzcEtRgAph2kovUw1.S8BGGeDY%2FBPyzGREiju9HyvhFXQrk%2FPq1MeAVQetHXA; Path=/; HttpOnly
< Vary: Origin
< Www-Authenticate: Basic realm="Users"
< Content-Length: 12
< Content-Type: text/plain; charset=utf-8
<
* Curl_http_done: called premature == 0
* Connection #0 to host 127.0.0.1 left intact
Unauthorized

#### Auth fail, without Origin header set:
$ curl -v -H "Host:whoami.docker.localhost" 127.0.0.1
* Rebuilt URL to: 127.0.0.1/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host:whoami.docker.localhost
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Access-Control-Allow-Credentials: true
< Date: Thu, 15 Oct 2020 20:08:07 GMT
< Set-Cookie: sails.sid=s%3AJEZ7M0ukFkiN8Vr7HJbpOVCj3zjGVZHc.SHcD2AQxtyTP2SK%2BDiMEr0x8bStaPNh9%2BActFBrUfsc; Path=/; HttpOnly
< Vary: Origin
< Www-Authenticate: Basic realm="Users"
< Content-Length: 12
< Content-Type: text/plain; charset=utf-8
<
* Curl_http_done: called premature == 0
* Connection #0 to host 127.0.0.1 left intact
Unauthorized

#### Auth success:
$ curl -v -H "Origin:http://mylaptop" -H "Authorization: Basic cG9zdG1hbjpwYXNzd29yZA==" -H "Host:whoami.docker.localhost" 127.0.0.1
* Rebuilt URL to: 127.0.0.1/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host:whoami.docker.localhost
> User-Agent: curl/7.52.1
> Accept: */*
> Origin:http://mylaptop
> Authorization: Basic cG9zdG1hbjpwYXNzd29yZA==
>
< HTTP/1.1 200 OK
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Origin: http://mylaptop
< Content-Length: 439
< Content-Type: text/plain; charset=utf-8
< Date: Thu, 15 Oct 2020 20:09:45 GMT
< Vary: Origin
<
Hostname: fd9bf79097b4
IP: 127.0.0.1
IP: 172.27.0.2
RemoteAddr: 172.27.0.3:58820
GET / HTTP/1.1
Host: whoami.docker.localhost
User-Agent: curl/7.52.1
Accept: */*
Accept-Encoding: gzip
Authorization: Basic cG9zdG1hbjpwYXNzd29yZA==
Origin: http://mylaptop
X-Forwarded-For: 172.27.0.1
X-Forwarded-Host: whoami.docker.localhost
X-Forwarded-Port: 80
X-Forwarded-Proto: http
X-Forwarded-Server: b6a52fb6a617
X-Real-Ip: 172.27.0.1

* Curl_http_done: called premature == 0
* Connection #0 to host 127.0.0.1 left intact
```

Do you have a reproducible request example and the output of its response (with missing headers)?

**Comment by letian0805** [comment]
@ddtmachado  CORS headers must be in front of ForwardAuth. If CORS is in behind of ForwardAuth and authorization failed, CORS will failed.

**Comment by ddtmachado** [comment]
@letian0805 I guess it already is in front -> `traefik.http.routers.backend-app.middlewares=cors@docker,auth@docker`
also there is output of a failed authentication with CORS on the examples I showed

**Comment by alewkinr** [comment]
Hello there, have absolutely the same issue. Check it out:
```
kind: Deployment
apiVersion: apps/v1
metadata:
  namespace: default
  name: traefik
  labels:
    app: traefik

spec:
  replicas: 1
  selector:
    matchLabels:
      app: traefik
  template:
    metadata:
      labels:
        app: traefik
    spec:
      serviceAccountName: traefik
      containers:
        - name: traefik
          image: traefik:v2.2
          args:
            - --api.dashboard=true
            - --api.insecure
            - --accesslog
            - --entrypoints.web.Address=:80
            - --entrypoints.websecure.Address=:443
            - --entrypoints.web.http.redirections.entryPoint.to=websecure
            - --entrypoints.web.http.redirections.entryPoint.scheme=https
            - --providers.kubernetescrd
            - --certificatesresolvers.myresolver.acme.tlschallenge=true
            - --certificatesresolvers.myresolver.acme.email=example@example.com
            - --certificatesresolvers.myresolver.acme.storage=acme.json
            - --certificatesresolvers.myresolver.acme.caserver=https://acme-v02.api.letsencrypt.org/directory
          ports:
            - name: web
              containerPort: 80
            - name: websecure
              containerPort: 443
            - name: admin
              containerPort: 8080
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: cors-header
  namespace: default
spec:
  headers:
    accessControlAllowMethods:
      - "POST"
      - "DELETE"
      - "GET"
      - "OPTIONS"
      - "PUT"
      - "PATCH"
    accessControlAllowHeaders:
      - "Content-Type"
      - "Authorization"
      - "*"
    accessControlAllowOriginList:
      - "..."
      - "..."
      - "..."
    accessControlMaxAge: 240
    accessControlAllowCredentials: true
    addVaryHeader: true
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: auth
  namespace: default
spec:
  forwardAuth:
    address: http://auth:8080/v1/validate
    authResponseHeaders:
      - "..."
      - "..."
---
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: client
  namespace: default
  resourceVersion: "46808449"
  selfLink: /apis/traefik.containo.us/v1alpha1/namespaces/default/ingressroutes/client
  uid: ...
spec:
  entryPoints:
  - websecure
  routes:
  - kind: Rule
    match: Host(`.....`) && PathPrefix(`/client`)
    middlewares:
    - name: redirectscheme
    - name: cors-header
    - name: auth
    - name: stripprefix
    services:
    - name: client
      port: 8080
  tls:
    certResolver: myresolver
```

As you can see, my cars mw is before auth, but the pipeline of middleware fails. Here is curl log:
```
> GET /client/v1/clients/3/status HTTP/2
> Host: ....
> authorization: Bearer myAuthToken
> accept: */*

* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!

< HTTP/2 401 
< content-type: application/json
< date: Thu, 29 Oct 2020 23:26:54 GMT
< x-content-type-options: nosniff
< content-length: 27
```
Traefik logs:
```
79.139.190.172 - - [29/Oct/2020:23:26:54 +0000] "GET /client/v1/clients/3/status HTTP/2.0" 401 27 "-" "-" 119270 "default-client-59c8cfc0a88ad80610d0@kubernetescrd" "-" 2ms
```

Any ways to fix it, except setting up CORS in auth application?

**Comment by rtribotte** [comment]
Hello @letian0805 @vd2org,

Can you provide a minimal reproducible case showing that the order `CORS -> ForwardAuth` breaks CORS in case of auth failure ? 
Can you try with the v2.3 ?

**Comment by letian0805** [comment]
@rtribotte 
 The order `ForwardAuth -> CORS` will breaks CORS in case of auth failure,  not `CORS -> ForwardAuth`.

**Comment by rtribotte** [comment]
Hello @letian0805,

I fear that this is the expected behavior. 
If the order is  `ForwardAuth -> CORS`, the `Headers` middleware is not reached when the authentication fails.

**Comment by alewkinr** [comment]
@rtribotte I have issue even with case `CORS —> ForwardAuth` 
I saw in docs something like that:
```
The ForwardAuth middleware delegate the authentication to an external service. If the service response code is 2XX, access is granted and the original request is performed. Otherwise, the response from the authentication server is returned.
```

I guess the returned response from authentication service overwrites traffic response.

**Comment by rtribotte** [comment]
Hello @alewkinr,

Which version do you use ? Since the v2.2.10 this should have been fixed.

**Comment by alewkinr** [comment]
@rtribotte here's my Traefik version, checked from CLI in k8s pod

```
/ # traefik version
Version:      2.2.1
Codename:     chevrotin
Go version:   go1.14.2
Built:        2020-04-29T18:02:09Z
OS/Arch:      linux/amd64
```

**Comment by rtribotte** [comment]
@alewkinr You can look at #7230, this is the PR that fixes your issue, and as previously said, this fix has been shipped with the v2.2.10, so you currently don't benefit of it.

**Comment by vd2org** [comment]
Seems to be fixed in 2.3, but I'm still testing.

**Comment by alewkinr** [comment]
> @alewkinr You can look at #7230, this is the PR that fixes your issue, and as previously said, this fix has been shipped with the v2.2.10, so you currently don't benefit of it.

Thx, it's fixed in 2.2.10

**Comment by vd2org** [comment]
Yes! Works! Thank you guys!
